srtcore/socketconfig.h

```diff
@@ -390,6 +390,39 @@ inline bool cast_optval(const void* optval, int optlen)
     return false;
 }
 
+/// @brief Sets optval and optlen to the value provided and its type size
+/// @tparam T value type (int, bool, etc.)
+/// @param[in,out] optval will be set to the value provided
+/// @param[in,out] optlen the length of the optval, will be set according to type T
+/// @param[in] value The value to be set in optval 
+template <typename T>
+inline void cast_set_optval(void* optval, int& optlen, const T& value)
+{
+    *reinterpret_cast<T*>(optval) = value;
+    optlen = sizeof(T);
+}
+
+template <>
+inline void cast_set_optval(void* optval, int& optlen, const bool& value)
+{
+    using namespace srt_logging;
+    if (optlen == 0 || optlen == sizeof(bool))
+    {
+        *reinterpret_cast<bool*>(optval) = value;
+        optlen = sizeof(bool);
+    }
+    else if (optlen >= sizeof(int))
+    {
+        *reinterpret_cast<int*>(optval) = value;
+        optlen = sizeof(int);
+    }
+    else
```

Comment:
First, when optlen is negative, `optlen >= sizeof(int)` is true due to cast int to unsigned, so the check doesn't work as expected.

I cannot find in the doc which specifies `optlen` is value-result parameter in the case of `getsockopt` for bool type other than "similar to the system `setsockopt/getsockopt` functions". The doc hints the other way that `optlen` is ignored by the implementation:

- `bool` - Requires the use of a boolean type (`<stdbool.h>` for C, or built-in
for C++). When *setting* an option, passing the value through an `int` type is
also properly recognized. When *getting* an option, however, you should use the
`bool` type, although you can risk passing a variable of `int` type initialized
with 0 and then checking if the resulting value is equal to 0 (just don't compare
the result with 1).

So it may break some code which doesn't initialize `optlen`.

In my opinion, the doc should be updated before change the behavior. Or just ignore the value of optlen for bool/int types.

Reply:
Here is an example.
[gst-plugins-bad](https://gitlab.freedesktop.org/gstreamer/gst-plugins-bad/-/merge_requests/2149/diffs)

gstreamer srt plugin use incorrect optlen in
```
    if (srt_getsockflag (caller->sock, SRTO_PAYLOADSIZE, &payload_size,
            &optlen)) {
```
Since the value of optlen is ignored by libsrt, the bug in gst srt plugin doesn't reveal.

Reply:
Yes, it definitely does not help that the [documentation](https://github.com/Haivision/srt/blob/master/docs/API/API-functions.md#srt_getsockopt) does not mention that `*optlen` should be set to anything in particular before calling `srt_getsockopt()`/`srt_getsockflag()` functions so it can't even be called a proper gst srt plugin bug. I don't think the contents of `*optlen` can be relied on at this point, at least not for a minor version release where API breakage is verboten. Maybe this change should be postponed until a bigger release?

Reply:
Maybe this way then. At least until it is asked in docs to provide a value of `optlen`.

```c++
template <>
inline void cast_set_optval(void* optval, int& optlen, const bool& value)
{
    using namespace srt_logging;
    if (optlen == sizeof(int))
    {
        *reinterpret_cast<int*>(optval) = value;
        optlen = sizeof(int);
    }
    else
    {
        *reinterpret_cast<bool*>(optval) = value;
        optlen = sizeof(bool);
    }
}
```